pkg/kubelet/kuberuntime/kuberuntime_image.go

```diff
@@ -32,14 +34,25 @@ import (
 
 // PullImage pulls an image from the network to local storage using the supplied
 // secrets if necessary.
-func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecontainer.ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig) (string, error) {
+func (m *kubeGenericRuntimeManager) PullImage(ctx context.Context, image kubecontainer.ImageSpec, pullSecrets []v1.Secret, podSandboxConfig *runtimeapi.PodSandboxConfig, serviceAccountName string) (string, error) {
 	img := image.Image
 	repoToPull, _, _, err := parsers.ParseImageName(img)
 	if err != nil {
 		return "", err
 	}
 
-	keyring, err := credentialprovidersecrets.MakeDockerKeyring(pullSecrets, m.keyring)
+	// construct the dynamic keyring using the providers we have in the kubelet
+	var podName, podNamespace, podUID string
+	if utilfeature.DefaultFeatureGate.Enabled(features.KubeletServiceAccountTokenForCredentialProviders) {
+		if podSandboxConfig != nil && podSandboxConfig.Metadata != nil {
+			podName = podSandboxConfig.Metadata.Name
+			podNamespace = podSandboxConfig.Metadata.Namespace
+			podUID = podSandboxConfig.Metadata.Uid
+		}
+	}
+	externalCredentialProviderKeyring := credentialproviderplugin.NewExternalCredentialProviderDockerKeyring(podName, podNamespace, podUID, serviceAccountName)
```

Comment:
I don't follow the gate off use case here passing in nil for the pod fields.. and continuing to the NewExternalCredentialProviderDockerKeyring func which will log and append with the nil detail each additional keyring provider.. then append below.

I don't follow the reason for the gate check just being for whether or not the additional keys have pod meta..

Reply:
The feature gate check here is just to ensure no additional metadata is passed to the plugins when the feature isn't enabled. The construction of the plugin is slightly different compared to before but when feature gate is disabled (1) the plugin will not get any of the additional metadata (2) feature gate checks in the plugin code ensure no change in behavior. The checks are at all layers to ensure we don't break any wiring when the feature gate is disabled.

Reply:
kk then suggest changing / eliminating the log when the gate is off and possibly checking to make sure no keys are injected?

```
klog.V(4).InfoS("Generating per pod credential provider", "provider", key, "podName", podName, "podNamespace", podNamespace, "podUID", podUID, "serviceAccountName", serviceAccountName)
```

 to indicate the difference

Reply:
trust but verify.. I expect this code path to be modified by cloud providers..

Reply:
> I expect this code path to be modified by cloud providers.. and it reads to me like we continue to union the response even if the return is supposed to be nil.

The plugin provider is never nil. Previously we registered the plugins at startup and constructed a keyring that's static. Now we are constructing the keyring dynamically as part of image pull request with some optional metadata. `perPodPluginProvider` is wrapping the plugin provider with additional fields and this shouldn't have any impact on cloud providers using the old plugin. Cloud providers will need to update the plugin if they want to use it with service account mode because there will be a new `ServiceAccountToken` in the `CredentialProviderRequest`.

Reply:
> kk then suggest changing / eliminating the log when the gate is off and possibly checking to make sure no keys are injected?

Agreed on modifying the log based on feature gate enabled. I'll make the change. We do have validation in the plugin to not use the metadata but I can configure the `perPodPluginProvider` with empty metadata if the feature flag is disabled as another layer of sanity check.

Reply:
Updated it! LMK what you think

Reply:
thanks for unwinding that for me..

Reply:
looks good thread closed